File: tauri/src-tauri/src/core/tools/search_files.rs
use async_trait::async_trait;
use std::collections::HashMap;

use super::{Tool, ToolCategory, ToolExample, resolve_path, truncate_result};

pub struct SearchFilesTool;

#[async_trait]
impl Tool for SearchFilesTool {
    fn name(&self) -> String { "search_files".to_string() }

    fn description(&self) -> String {
        "Search for a text pattern across files in the workspace using recursive grep. \
         Returns matching lines with file paths and line numbers. Use this to find function \
         definitions, usages, imports, error messages, or any text pattern. \
         Automatically skips hidden files and common build directories (node_modules, target, .git)."
            .to_string()
    }

    fn parameters(&self) -> String {
        serde_json::json!({
            "type": "object",
            "properties": {
                "pattern": {
                    "type": "string",
                    "description": "The text pattern to search for (plain text or regex)."
                },
                "path": {
                    "type": "string",
                    "description": "Optional subdirectory to search in (relative to workspace). Defaults to entire workspace."
                },
                "include_glob": {
                    "type": "string",
                    "description": "Optional glob to filter files (e.g. '*.rs', '*.ts'). Defaults to all files."
                }
            },
            "required": ["pattern"]
        })
        .to_string()
    }

    fn category(&self) -> ToolCategory { ToolCategory::CodeIntelligence }

    fn examples(&self) -> Vec<ToolExample> {
        vec![ToolExample {
            description: "Find all usages of a function".to_string(),
            xml: "<tool_call>\n  <name>search_files</name>\n  <parameters>\n    <pattern>fn create_instance</pattern>\n    <include_glob>*.rs</include_glob>\n  </parameters>\n</tool_call>".to_string(),
        }]
    }

    async fn execute(&self, params: HashMap<String, String>, workspace: &str) -> Result<String, String> {
        let pattern = params.get("pattern").ok_or("Missing 'pattern' parameter")?;
        let sub_path = params.get("path").map(|s| s.as_str()).unwrap_or(".");
        let search_dir = resolve_path(workspace, sub_path);

        let mut cmd_parts = vec![
            "grep".to_string(),
            "-rn".to_string(),
            "--color=never".to_string(),
            "--exclude-dir=.git".to_string(),
            "--exclude-dir=node_modules".to_string(),
            "--exclude-dir=target".to_string(),
            "--exclude-dir=.nuxt".to_string(),
            "--exclude-dir=.output".to_string(),
            "--exclude-dir=dist".to_string(),
        ];

        if let Some(glob) = params.get("include_glob") {
            cmd_parts.push(format!("--include={}", glob));
        }

        cmd_parts.push("--".to_string());
        cmd_parts.push(pattern.clone());
        cmd_parts.push(search_dir.to_string_lossy().to_string());

        let output = tokio::process::Command::new("grep")
            .args(&cmd_parts[1..])
            .output()
            .await
            .map_err(|e| format!("Failed to run grep: {}", e))?;

        let stdout = String::from_utf8_lossy(&output.stdout);

        if stdout.is_empty() {
            return Ok(format!("No matches found for '{}' in {}", pattern, search_dir.display()));
        }

        let match_count = stdout.lines().count();
        let header = format!("Found {} matches for '{}':\n\n", match_count, pattern);

        Ok(truncate_result(&format!("{}{}", header, stdout), 15_000))
    }
}

File: electron/electron/core/tools/search_files.ts
import * as fs from 'fs'
import * as path from 'path'
import type { Tool, ToolExample } from './types'
import { ToolCategory } from './types'
import { resolvePath, truncateResult, globToRegex } from './utils'

class SearchFilesTool implements Tool {
    name() { return 'search_files' }
    description() { return 'Search for a pattern in all files recursively. Returns matching lines with file paths and line numbers.' }
    category() { return ToolCategory.CodeIntelligence }
    isDestructive() { return false }
    parameters() {
        return JSON.stringify({
            properties: {
                pattern: { type: 'string', description: 'Text or regex pattern to search for' },
                path: { type: 'string', description: 'Directory to search in (default: workspace root)' },
                file_pattern: { type: 'string', description: 'Glob pattern to filter files, e.g. "*.ts"' },
            },
            required: ['pattern'],
        })
    }
    examples(): ToolExample[] {
        return [{ description: 'Find all TODO comments', xml: '<tool_call>\n{"name": "search_files", "arguments": {"pattern": "TODO"}}\n</tool_call>' }]
    }
    async execute(params: Record<string, string>, workspace: string): Promise<string> {
        const searchPath = resolvePath(params.path || '.', workspace)
        const pattern = params.pattern || ''
        const excludeDirs = ['node_modules', '.git', 'target', 'dist', '.nuxt', '.output', '__pycache__']
        const results: string[] = []

        const searchDir = (dir: string) => {
            if (results.length > 100) return
            let entries: fs.Dirent[]
            try { entries = fs.readdirSync(dir, { withFileTypes: true }) } catch { return }

            for (const entry of entries) {
                if (results.length > 100) break
                if (excludeDirs.includes(entry.name)) continue
                const full = path.join(dir, entry.name)

                if (entry.isDirectory()) {
                    searchDir(full)
                } else if (entry.isFile()) {
                    if (params.file_pattern && !entry.name.match(globToRegex(params.file_pattern))) continue
                    try {
                        const content = fs.readFileSync(full, 'utf-8')
                        const lines = content.split('\n')
                        for (let i = 0; i < lines.length; i++) {
                            if (lines[i].includes(pattern)) {
                                const rel = path.relative(workspace, full)
                                results.push(`${rel}:${i + 1}: ${lines[i].trim()}`)
                            }
                        }
                    } catch { /* binary or unreadable */ }
                }
            }
        }

        searchDir(searchPath)
        if (results.length === 0) return `No matches found for "${pattern}"`
        return truncateResult(results.join('\n'))
    }
}

export default SearchFilesTool
